keps/sig-cli/3895-kubectl-delete-interactivity/README.md

```diff
@@ -0,0 +1,926 @@
+
+### Goals
+
+- Introduce interactive(-i) flag to be used for asking confirmation to user to proceed actual delete operation or not.
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+- Smart solutions on client side to detect deletion is accidental or not
+- Deletion protection mechanism on server side
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+When a user passes `--interactive`/`-i` flag to `kubectl delete` command, all the resources planned to be deleted will be 
+shown to the user as preview. Command continues deletion process if user confirms by typing `y`. If user types `n` or
+any other value other than `y`, process will be cancelled and command returns a message with zero exit code.
+
+```sh
+$ kubectl delete --interactive -f test_deployment.yaml 
+You are about to delete the following 2 resource(s):
+Deployment/dockerd
+Deployment/dockerd2
+Do you want to continue? (y/n): n
+deletion is cancelled
+
+$ kubectl delete -i -f test_deployment.yaml
+You are about to delete the following 2 resource(s):
+Deployment/dockerd
+Deployment/dockerd2
+Do you want to continue? (y/n): y
+deployment.apps "dockerd" deleted
+deployment.apps "dockerd2" deleted
+```
+
+For the alpha phase, this flag will be hidden behind `KUBECTL_INTERACTIVE_DELETE` environment variable and
+can only be use when this `KUBECTL_INTERACTIVE_DELETE=true` environment variable is exported.
+
+### Implementation Details/Design
+
+One of major advantage of this feature is the ability to preview all resources before deletion. That means
+we need to traverse all resources twice. First one is for display, and the other one is for deletion.
+However, that comes along with a technical limitation that `Visit` function does not support traversing twice
+as stated in [here](https://github.com/kubernetes/kubernetes/blob/6d7a9048b67d81e57923892650a23636f1afba42/staging/src/k8s.io/cli-runtime/pkg/resource/result.go#L91-L94).
+But we also cannot use `Infos` function as suggested in the link above, because that unties the builder configuration
+and deletion logic.
+
+In order to overcome this limitation, this feature will use separate builder to show resources as preview. Current builder
+will continue being used as deletion traverse, other builder will be used for preview. Builders must match in terms of 
+configurations to prevent such discrepancies and required tests will be added to assure that configurations are same or not.
+Moreover, deletion will only be performed to the resources whose `uidMap` match the ones in the preview list.
```

Comment:
```suggestion
Moreover, deletion will only be performed on the resources whose `.metadata.uid` match the ones gathered in the preview list, to ensure we remove only the resources a user explicitly agreed to. 
```

Reply:
Does this need a stronger guarantee like RV based matching?

Reply:
Thanks for review @enj.
We need to not delete the resources whose are not in the preview list(especially the cases where user types `y` after a couple of hours and preview list becomes stale most likely). And `uid` based check allows us to do it and this looks sufficient for our aim(Since `kubectl delete` does not know resource versions of resources, that might complicate the issue).